services/src/main/java/org/keycloak/authentication/requiredactions/DeleteAccount.java

```diff
@@ -0,0 +1,166 @@
+
+import org.jboss.logging.Logger;
+import org.keycloak.Config;
+import org.keycloak.authentication.InitiatedActionSupport;
+import org.keycloak.authentication.RequiredActionContext;
+import org.keycloak.authentication.RequiredActionFactory;
+import org.keycloak.authentication.RequiredActionProvider;
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.forms.login.freemarker.model.UrlBean;
+import org.keycloak.models.AccountRoles;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.Constants;
+import org.keycloak.models.KeycloakContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserManager;
+import org.keycloak.models.UserModel;
+import org.keycloak.services.ForbiddenException;
+import org.keycloak.services.messages.Messages;
+import org.keycloak.theme.Theme;
+import org.keycloak.utils.UserHelper;
+
+public class DeleteAccount implements RequiredActionProvider, RequiredActionFactory {
+
+  public static final String PROVIDER_ID = "delete_account";
+
+  private static final Logger logger = Logger.getLogger(DeleteAccount.class);
+
+  @Override
+  public String getDisplayText() {
+    return "Delete Account";
+  }
+
+  @Override
+  public void evaluateTriggers(RequiredActionContext context) {
+
+  }
+
+  @Override
+  public void requiredActionChallenge(RequiredActionContext context) {
+    UserModel user = context.getAuthenticationSession().getAuthenticatedUser();
+    RealmModel realm = context.getRealm();
+
+    if(!UserHelper.isDeleteAccountAllowed(realm, user)) {
+      throw new ForbiddenException();
+    }
+
+    context.challenge(context.form().createForm("delete-account-confirm.ftl"));
+  }
+
+
+  @Override
+  public void processAction(RequiredActionContext context) {
+    KeycloakSession session = context.getSession();
+    EventBuilder eventBuilder = context.getEvent();
+    KeycloakContext keycloakContext = session.getContext();
+    RealmModel realm = keycloakContext.getRealm();
+    UserModel user = keycloakContext.getAuthenticationSession().getAuthenticatedUser();
+
+    try {
+      if(!UserHelper.isDeleteAccountAllowed(realm, user)) {
+        throw new ForbiddenException();
+      }
+      boolean removed = new UserManager(session).removeUser(realm, user);
+
+      if (removed) {
+        eventBuilder.event(EventType.DELETE_ACCOUNT)
+            .client(keycloakContext.getClient())
+            .user(user)
+            .detail(Details.USERNAME, user.getUsername())
+            .success();
+      } else {
+        eventBuilder.event(EventType.DELETE_ACCOUNT)
+            .client(keycloakContext.getClient())
+            .user(user)
+            .detail(Details.USERNAME, user.getUsername())
+            .error("User could not be deleted");
+        context.failure();
```

Comment:
This code needs updating as it needs to handle the case that the authentication flow can now not succeed. Not quite sure what it needs to do, but it should make sure the session is cleared, probably reset the authentication flow, set no user in the authentication flow, etc..

@mposolda Do you have any tips on how we can handle this?

Reply:
+1

In case of the authentication flows, the reset of the flow can be done with the usage of method `AuthenticationFlowContext.resetFlow()` or eventually the `AuthenticationFlowContext.resetFlow(Runnable afterResetListener)`, which allows to trigger some code after the reset of the authentication session is done. This will do all the mentioned things (cleanup user from the authSession, cleanup the state of the authentication flow etc). See the `AuthenticationProcessor.resetFlow` to see the details of everything, which is done.

It seems that RequiredActionContext does not yet have this possibility to reset the flow, so the new method may need to be added to RequiredActionContext.

It will be ideal if actual "implementation" of the reset flow (currently `AuthenticationProcessor.resetFlow`) is somehow shared for both authenticationFlowContext and RequiredActionContext, so that there is not same duplicated code for this implementation. Maybe just some static method, which can be called from both places will do the trick - or maybe there is some even more elegant solution.

Reply:
Thanks for the input. I found a way to reset the flow/ clean the session using the current `RequiredActionContext` without adding new methods: 

```
context.getAuthenticationSession().removeRequiredAction(PROVIDER_ID);
    context.getAuthenticationSession().removeAuthNote(AuthenticationProcessor.CURRENT_AUTHENTICATION_EXECUTION);
  AuthenticationManager.setKcActionStatus(PROVIDER_ID, status, context.getAuthenticationSession());
```

The status can be either: `KcActionStatus.SUCCESS` and `KcActionStatus.ERROR`. 

In the case of a success, I just send a challenge to an `info.ftl` page with a message that says the user has been deleted successfully. The info page has bottom links to login page. 

Otherwise, I stay on the same confirm page and display an error message.